docarray/index/backends/qdrant.py

```diff
@@ -0,0 +1,478 @@
+                for column_name, column_info in self._column_infos.items()
+                if column_info.db_type == 'vector'
+            }
+            self._client.create_collection(
+                collection_name=self._db_config.collection_name,
+                vectors_config=vectors_config,
+                shard_number=self._db_config.shard_number,
+                replication_factor=self._db_config.replication_factor,
+                write_consistency_factor=self._db_config.write_consistency_factor,
+                on_disk_payload=self._db_config.on_disk_payload,
+                hnsw_config=self._db_config.hnsw_config,
+                optimizers_config=self._db_config.optimizers_config,
+                wal_config=self._db_config.wal_config,
+                quantization_config=self._db_config.quantization_config,
+            )
+            self._client.create_payload_index(
+                collection_name=self._db_config.collection_name,
+                field_name='__generated_vectors',
+                field_schema=rest.PayloadSchemaType.KEYWORD,
+            )
+
+    def _index(self, column_to_data: Dict[str, Generator[Any, None, None]]):
+        rows = self._transpose_col_value_dict(column_to_data)
+        # TODO: add batching the documents to avoid timeouts
+        points = [self._build_point_from_row(row) for row in rows]
+        self._client.upsert(
+            collection_name=self._db_config.collection_name,
+            points=points,
+        )
+
+    def num_docs(self) -> int:
+        return self._client.count(collection_name=self._db_config.collection_name).count
+
+    def _del_items(self, doc_ids: Sequence[str]):
+        items = self._get_items(doc_ids)
+        if len(items) < len(doc_ids):
+            found_keys = set(item['id'] for item in items)  # type: ignore[index]
+            missing_keys = set(doc_ids) - found_keys
+            raise KeyError('Document keys could not found: %s' % ','.join(missing_keys))
+
+        self._client.delete(
+            collection_name=self._db_config.collection_name,
+            points_selector=rest.PointIdsList(
+                points=[self._to_qdrant_id(doc_id) for doc_id in doc_ids],
+            ),
+        )
+
+    def _get_items(
+        self, doc_ids: Sequence[str]
+    ) -> Union[Sequence[TSchema], Sequence[Dict[str, Any]]]:
+        response, _ = self._client.scroll(
+            collection_name=self._db_config.collection_name,
+            scroll_filter=rest.Filter(
+                must=[
+                    rest.HasIdCondition(
+                        has_id=[self._to_qdrant_id(doc_id) for doc_id in doc_ids],
+                    ),
+                ],
+            ),
+            limit=len(doc_ids),
+            with_payload=True,
+            with_vectors=True,
+        )
+        return [self._convert_to_doc(point) for point in response]
+
+    def execute_query(self, query: Query, *args, **kwargs) -> DocList:
```

Comment:
here the users should also be able to pass in a raw query that does not come from our query builder, i.e. a python dict or string like this (copied from your docs):

```python
{
    "filter": {
        "must": [
            {
                "key": "city",
                "match": {
                    "value": "London"
                }
            }
        ]
    },
    "params": {
        "hnsw_ef": 128,
        "exact": false
    },
    "vector": [0.2, 0.1, 0.9, 0.7],
    "limit": 3
}
```

This is meant as a fallback option for any functionality that may not be covered by the DocArray API, we don't want to lock away Qdrant functionality from our users. Is that possible without too much hassle?

Reply:
I'm still working on that one - I'd like to avoid sending raw HTTP requests, and we need to do some conversion.

Reply:
@JohannesMessner This is already implemented, so raw queries might be passed to `.execute_query`